File: apps/api/src/config/flags.ts
import type { Prisma } from "@prisma/client";
import { prisma } from "../lib/prisma";
import { connectRedis, redis } from "../lib/redis";

/**
 * Runtime feature flags — the same pattern as config/resolve.ts (Redis-cached
 * DB rows, busted on write, audited in ConfigAudit), for booleans the dashboard
 * flips rather than numeric risk fields.
 *
 * Resolution: DB row ► env default ► false. So a flag is off until someone
 * either sets the env var or flips it in the UI, and the UI always wins.
 */

/** Record every raw strategy candidate (protection layers OFF, observe-only). */
export const RAW_FEED_FLAG = "raw_signal_feed";

/** Env var consulted when a flag has no DB row yet. */
const ENV_DEFAULTS: Record<string, string> = {
  [RAW_FEED_FLAG]: "RAW_SIGNAL_FEED",
};

const FLAG_CACHE_KEY = "config:flags:rows";
const CACHE_TTL_S = 300; // safety net; writes bust the cache explicitly

export type FlagSource = "db" | "env" | "default";

export interface FlagState {
  key: string;
  enabled: boolean;
  source: FlagSource;
}

type FlagRow = { key: string; enabled: boolean };

function envDefault(key: string): boolean | null {
  const envName = ENV_DEFAULTS[key];
  if (!envName) return null;
  const raw = process.env[envName];
  if (raw == null || raw.trim() === "") return null;
  return ["1", "true", "yes", "on"].includes(raw.trim().toLowerCase());
}

async function getFlagRows(): Promise<FlagRow[]> {
  try {
    await connectRedis();
    const cached = await redis.get(FLAG_CACHE_KEY);
    if (cached) return JSON.parse(cached) as FlagRow[];
  } catch {
    /* fall through to DB */
  }
  const rows = await prisma.featureFlag.findMany({ select: { key: true, enabled: true } });
  try {
    await redis.set(FLAG_CACHE_KEY, JSON.stringify(rows), "EX", CACHE_TTL_S);
  } catch {
    /* cache is best-effort */
  }
  return rows;
}

/** Invalidate the flag cache — call after any flag write. */
export async function bustFlagCache(): Promise<void> {
  try {
    await connectRedis();
    await redis.del(FLAG_CACHE_KEY);
  } catch {
    /* best-effort */
  }
}

export async function getFlag(key: string): Promise<FlagState> {
  const row = (await getFlagRows()).find((r) => r.key === key);
  if (row) return { key, enabled: row.enabled, source: "db" };
  const env = envDefault(key);
  if (env != null) return { key, enabled: env, source: "env" };
  return { key, enabled: false, source: "default" };
}

/**
 * Is a flag on? Never throws: an unreachable DB/Redis resolves to the env
 * default (then false), so a flag outage can only turn a feature OFF.
 */
export async function isFlagEnabled(key: string): Promise<boolean> {
  try {
    return (await getFlag(key)).enabled;
  } catch (err) {
    console.error("[flags] resolve failed:", err instanceof Error ? err.message : err);
    return envDefault(key) ?? false;
  }
}

/** Convenience for the hot path in the signal gate. */
export function isRawFeedEnabled(): Promise<boolean> {
  return isFlagEnabled(RAW_FEED_FLAG);
}

export async function setFlag(actor: string, key: string, enabled: boolean): Promise<FlagState> {
  const before = await prisma.featureFlag.findUnique({ where: { key } });
  const after = await prisma.featureFlag.upsert({
    where: { key },
    create: { key, enabled },
    update: { enabled },
  });

  try {
    await prisma.configAudit.create({
      data: {
        actor,
        entity: "FeatureFlag",
        scope: "GLOBAL",
        scopeKey: key,
        before: (before ?? {}) as unknown as Prisma.InputJsonValue,
        after: after as unknown as Prisma.InputJsonValue,
      },
    });
  } catch (err) {
    console.error("[flags] audit write failed:", err instanceof Error ? err.message : err);
  }

  await bustFlagCache();
  return { key, enabled: after.enabled, source: "db" };
}

File: services/backend/app/domain/config/flags.py
"""Runtime feature flags — port of `apps/api/src/config/flags.ts`.

Same pattern as the risk config: Redis-cached DB rows, busted on write, audited
in `ConfigAudit`. Resolution is DB row ► env default ► false, so a flag is off
until someone sets the env var or flips it in the UI, and the UI always wins.
"""
from __future__ import annotations

import json
import logging
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Literal

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from ...core.settings import get_settings
from ...db.models import FeatureFlag
from ...db.redis_client import cache_del, cache_get, cache_set
from .store import audit

log = logging.getLogger("backend.flags")

#: Record every raw strategy candidate (protection layers OFF, observe-only).
RAW_FEED_FLAG = "raw_signal_feed"

FLAG_CACHE_KEY = "config:flags:rows"
CACHE_TTL_S = 300  # safety net; writes bust the cache explicitly

FlagSource = Literal["db", "env", "default"]


@dataclass(slots=True)
class FlagState:
    key: str
    enabled: bool
    source: FlagSource

    def as_dict(self) -> dict[str, object]:
        return {"key": self.key, "enabled": self.enabled, "source": self.source}


def _env_default(key: str) -> bool | None:
    """The env var consulted when a flag has no DB row yet."""
    if key == RAW_FEED_FLAG:
        return get_settings().raw_signal_feed
    return None


async def _flag_rows(session: AsyncSession) -> list[dict[str, object]]:
    cached = await cache_get(FLAG_CACHE_KEY)
    if cached:
        try:
            return json.loads(cached)
        except json.JSONDecodeError:
            pass
    rows = (await session.execute(select(FeatureFlag.key, FeatureFlag.enabled))).all()
    payload = [{"key": r[0], "enabled": r[1]} for r in rows]
    await cache_set(FLAG_CACHE_KEY, json.dumps(payload), CACHE_TTL_S)
    return payload


async def bust_flag_cache() -> None:
    """Invalidate the flag cache — call after any flag write."""
    await cache_del(FLAG_CACHE_KEY)


async def get_flag(session: AsyncSession, key: str) -> FlagState:
    for row in await _flag_rows(session):
        if row["key"] == key:
            return FlagState(key, bool(row["enabled"]), "db")
    env = _env_default(key)
    if env is not None:
        return FlagState(key, env, "env")
    return FlagState(key, False, "default")


async def is_flag_enabled(session: AsyncSession, key: str) -> bool:
    """Is a flag on? Never raises — an outage can only turn a feature OFF."""
    try:
        return (await get_flag(session, key)).enabled
    except Exception as exc:
        log.error("[flags] resolve failed: %s", exc)
        return _env_default(key) or False


async def set_flag(session: AsyncSession, actor: str, key: str, enabled: bool) -> FlagState:
    """Upsert a flag, audit it, and bust the cache."""
    row = (
        await session.execute(select(FeatureFlag).where(FeatureFlag.key == key))
    ).scalar_one_or_none()
    before = {} if row is None else {"key": row.key, "enabled": row.enabled}

    if row is None:
        row = FeatureFlag(
            key=key, enabled=enabled, updatedAt=datetime.now(timezone.utc).replace(tzinfo=None)
        )
        session.add(row)
    else:
        row.enabled = enabled
        row.updatedAt = datetime.now(timezone.utc).replace(tzinfo=None)
    await session.flush()

    await audit(
        session,
        actor,
        "FeatureFlag",
        "GLOBAL",
        key,
        before,
        {"key": row.key, "enabled": row.enabled},
    )
    await session.commit()
    await bust_flag_cache()
    return FlagState(key, row.enabled, "db")
